File: leetcode/Algorithms/547.FriendCircles/Solution.py
class Solution(object):
    def findCircleNum(self, M):
        """
        :type M: List[List[int]]
        :rtype: int
        """
        result, length = 0, len(M)

        for i in xrange(length):
            if M[i][i] != 1:
                continue

            M[i][i] = -1
            queue = [i]
            while len(queue) != 0:
                index = queue.pop()
                for j in xrange(length):
                    if M[index][j] == 1 and M[j][j] == 1:
                        M[j][j] = -1
                        queue.append(j)
            result += 1

        return result

File: leetcode/Algorithms/547.FriendCircles/Solution.go
package main

func findCircleNum(M [][]int) int {
	tmp := make([]bool, len(M))
	result := 0
	for i := 0; i < len(M); i++ {
		if !tmp[i] {
			travel(M, i, tmp)
			result++
		}
	}
	return result
}

func travel(M [][]int, n int, tmp []bool) {
	tmp[n] = true
	for i := 0; i < len(M); i++ {
		if M[n][i] == 1 && !tmp[i] {
			travel(M, i, tmp)
		}
	}
}

